java/ql/src/semmle/code/java/dataflow/internal/ContainerFlow.qll

```diff
@@ -187,6 +187,24 @@ private predicate qualifierToMethodStep(Expr tracked, MethodAccess sink) {
   tracked = sink.getQualifier()
 }
 
+private predicate taintPreservingQualifierToFunctionalParameter(Method m, Parameter p) {
+  m.getDeclaringType() instanceof IterableType and
+  m.hasName("forEach")
+  or
+  m.(MapMethod).hasName("forEach")
+  or
+  m.getDeclaringType() instanceof IteratorType and
+  m.hasName("forEachRemaining")
+}
+
+private predicate qualifierToFunctionalParameterStep(Expr tracked, Parameter parameter) {
+  exists(MethodAccess ma |
+    taintPreservingQualifierToFunctionalParameter(ma.getMethod(), parameter) and
+    tracked = ma.getQualifier() and
```

Comment:
Maybe it would be good to use `DataFlow::getInstanceArgument(Call)` as suggested in https://github.com/github/codeql/pull/5814#discussion_r629889595